File: Main/src/main/java/com/redelf/commons/persistance/DataSerializer.java
package com.redelf.commons.persistance;

import android.text.TextUtils;

import com.google.firebase.crashlytics.FirebaseCrashlytics;
import com.google.gson.Gson;
import com.google.gson.JsonSyntaxException;
import com.redelf.commons.logging.Console;
import com.redelf.commons.persistance.base.Serializer;

import java.util.Iterator;
import java.util.List;
import java.util.Map;
import java.util.Set;

class DataSerializer implements Serializer {

    /*
        TODO: Create a flavor that uses Jackson lib for stream-like serialization / deserialization
    */
    private final Gson gsn;

    {

        gsn = new Gson();
    }

    @Override
    public <T> String serialize(byte[] cipherText, T originalGivenValue) {

        if (cipherText == null || cipherText.length == 0) {

            return null;
        }

        if (originalGivenValue == null) {

            return null;
        }

        Class<?> keyClassName = null;
        Class<?> valueClassName = null;

        char dataType;

        if (List.class.isAssignableFrom(originalGivenValue.getClass())) {

            List<?> list = (List<?>) originalGivenValue;
            if (!list.isEmpty()) {
                keyClassName = list.get(0).getClass();
            }
            dataType = DataInfo.TYPE_LIST;

        } else if (Map.class.isAssignableFrom(originalGivenValue.getClass())) {

            dataType = DataInfo.TYPE_MAP;
            Map<?, ?> map = (Map<?, ?>) originalGivenValue;

            if (!map.isEmpty()) {
                for (Map.Entry<?, ?> entry : map.entrySet()) {
                    keyClassName = entry.getKey().getClass();
                    valueClassName = entry.getValue().getClass();
                    break;
                }
            }

        } else if (Set.class.isAssignableFrom(originalGivenValue.getClass())) {

            Set<?> set = (Set<?>) originalGivenValue;
            if (!set.isEmpty()) {
                Iterator<?> iterator = set.iterator();
                if (iterator.hasNext()) {
                    keyClassName = iterator.next().getClass();
                }
            }
            dataType = DataInfo.TYPE_SET;

        } else {

            dataType = DataInfo.TYPE_OBJECT;
            keyClassName = originalGivenValue.getClass();
        }

        final DataInfo dataInfo = new DataInfo(

                cipherText,
                dataType,
                keyClassName != null ? keyClassName.getName() : null,
                valueClassName != null ? valueClassName.getName() : null,
                keyClassName,
                valueClassName
        );

        try {

            return gsn.toJson(dataInfo);

        } catch (OutOfMemoryError e) {

            Console.error(e);

            FirebaseCrashlytics.getInstance().recordException(e);

        } catch (Exception e) {

            Console.error(e);
        }

        return null;
    }

    @Override
    public DataInfo deserialize(String serializedText) {

        if (TextUtils.isEmpty(serializedText)) {

            return null;
        }

        try {

            final DataInfo dataInfo = gsn.fromJson(serializedText, DataInfo.class);

            if (dataInfo.getKeyClazzName() != null) {
                try {
                    dataInfo.setKeyClazz(Class.forName(dataInfo.getKeyClazzName()));
                } catch (ClassNotFoundException e) {
                    Console.error(e);
                }
            }

            if (dataInfo.getValueClazzName() != null) {
                try {
                    dataInfo.setValueClazz(Class.forName(dataInfo.getValueClazzName()));
                } catch (ClassNotFoundException e) {
                    Console.error(e);
                }
            }

            return dataInfo;

        } catch (JsonSyntaxException e) {

            Console.error(e);
        }

        return null;
    }
}

File: Main/src/main/java/com/redelf/commons/persistance/DataSerializer.kt
/*
 * Copyright (c) 2025 MeTube Share
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */


package com.redelf.commons.persistance

import com.google.gson.JsonSyntaxException
import com.redelf.commons.extensions.forClassName
import com.redelf.commons.extensions.isEmpty
import com.redelf.commons.extensions.recordException
import com.redelf.commons.logging.Console.error
import com.redelf.commons.obtain.Obtain
import com.redelf.commons.persistance.base.Parser
import com.redelf.commons.persistance.base.Serializer

internal class DataSerializer(private val parser: Obtain<Parser>) : Serializer {

    /*
        TODO: Create a flavor that uses Jackson lib for stream-like serialization / deserialization
    */

    override fun <T> serialize(cipherText: String?, value: T): String? {

        if (cipherText == null || cipherText.isEmpty()) {
            
            return null
        }

        if (value == null) {
            
            return null
        }

        var keyClassName: Class<*>? = null
        var valueClassName: Class<*>? = null

        var dataType: String

        if (MutableList::class.java.isAssignableFrom(value.javaClass)) {
            
            val list = value as MutableList<*>
            
            if (!list.isEmpty()) {

                keyClassName = list.get(0)?.javaClass
            }

            dataType = DataInfo.TYPE_LIST

        } else if (MutableMap::class.java.isAssignableFrom(value.javaClass)) {

            dataType = DataInfo.TYPE_MAP

            val map = value as MutableMap<*, *>

            if (!map.isEmpty()) {

                for (entry in map.entries) {

                    keyClassName = entry.key?.javaClass
                    valueClassName = entry.value?.javaClass
                    break
                }

            }

        } else if (MutableSet::class.java.isAssignableFrom(value.javaClass)) {

            val set = value as MutableSet<*>

            if (!set.isEmpty()) {

                val iterator: MutableIterator<*> = set.iterator()

                if (iterator.hasNext()) {

                    keyClassName = iterator.next()?.javaClass
                }
            }

            dataType = DataInfo.TYPE_SET

        } else {

            dataType = DataInfo.TYPE_OBJECT
            keyClassName = value.javaClass
        }

        val dataInfo = DataInfo(

            cipherText,
            dataType,
            keyClassName?.name,
            valueClassName?.name,
            keyClassName?.canonicalName?.forClassName(),
            valueClassName?.canonicalName?.forClassName()
        )

        try {

            return parser.obtain().toJson(dataInfo)

        } catch (e: OutOfMemoryError) {

            recordException(e)

        } catch (e: Throwable) {

            error(e)
        }

        return null
    }

    override fun deserialize(plainText: String?): DataInfo? {

        if (isEmpty(plainText)) {

            return null
        }

        try {

            val dataInfo = parser.obtain().fromJson<DataInfo?>(plainText, DataInfo::class.java)

            if (dataInfo?.keyClazzName != null) {

                dataInfo.keyClazz = dataInfo.keyClazzName?.forClassName()
            }

            if (dataInfo?.valueClazzName != null) {

                dataInfo.valueClazz = dataInfo.valueClazzName?.forClassName()
            }

            return dataInfo

        } catch (e: JsonSyntaxException) {

            error("Could not deserialize: $plainText")

            error(e)
        }

        return null
    }
}
